### PR TITLE
[beta backend] make dropbox config optional

### PR DESCRIPTION
**Pull request by gorzalczany**

### CODE REVIEW

File: service/src/main/kotlin/xyz/lebkuchenfm/api/xsounds/XSoundsEndpoint.kt [hunk 1]
```diff
@@ -57,8 +59,9 @@ fun Route.xSoundsRouting(xSoundsService: XSoundsService) {
 
             fileBytes?.let { bytes ->
                 // TODO: pass authenticated user name as "addedBy"
-                val sound = xSoundsService.addNewXSound(soundName, tags, bytes)
-                call.respond(HttpStatusCode.Created, sound.toResponse())
+                xSoundsService.addNewXSound(soundName, tags, bytes)
+                    .onSuccess { call.respond(HttpStatusCode.Created, it.toResponse()) }
+                    .onFailure { call.respond(HttpStatusCode.InternalServerError, "Could not create new x-sound.") }
```
Comment (Deseteral) [inline service/src/main/kotlin/xyz/lebkuchenfm/api/xsounds/XSoundsEndpoint.kt:64]: We should have standardized error JSON response on REST endpoints. Maybe we should put this one in the doc and do a pass on all endpoints later?

File: service/src/main/kotlin/xyz/lebkuchenfm/domain/xsounds/XSoundsService.kt [hunk 2]
```diff
@@ -13,11 +17,16 @@ class XSoundsService(private val repository: XSoundsRepository, private val file
...
 
-    suspend fun addNewXSound(soundName: String, tags: List<String>, bytes: ByteArray): XSound {
-        val fileUrl = fileRepository.uploadXSoundFile(soundName, bytes)
+    sealed interface NewSoundError {
+        data object CouldNotAddNewSound : NewSoundError
+    }
+
+    suspend fun addNewXSound(soundName: String, tags: List<String>, bytes: ByteArray): Result<XSound, NewSoundError> {
+        val fileUrl = fileRepository.uploadXSoundFile(soundName, bytes).get()
+            ?: return Err(NewSoundError.CouldNotAddNewSound)
```
Comment (Deseteral) [inline service/src/main/kotlin/xyz/lebkuchenfm/domain/xsounds/XSoundsService.kt:26]: This one should be about not being able to save the file into the file storage.
```suggestion
            ?: return Err(NewSoundError.CouldNotSaveFileToRepository)
```

File: service/src/main/kotlin/xyz/lebkuchenfm/domain/xsounds/XSoundsService.kt [hunk 3]
```diff
@@ -13,11 +17,16 @@ class XSoundsService(private val repository: XSoundsRepository, private val file
...
         val readySound = XSound(name = soundName, url = fileUrl, tags = tags, timesPlayed = 0, addedBy = null)
         repository.insert(readySound)
```
Comment (Deseteral) [inline service/src/main/kotlin/xyz/lebkuchenfm/domain/xsounds/XSoundsService.kt:28]: We should check if the sound with the same name already exists. This is another type of `NewSoundError`.

File: service/src/main/kotlin/xyz/lebkuchenfm/external/storage/dropbox/DropboxClient.kt [hunk 4]
```diff
@@ -29,21 +32,33 @@ import xyz.lebkuchenfm.external.storage.dropbox.models.DropboxFileUploadArgs
...
+            logger.warn { "Tried to use dropbox client but config is missing." }
+            return Err(DropboxClientError.ClientConfigMissing)
+        }
+
         if (bearerTokenStorage.isEmpty()) {
-            getInitialTokens()
+            try {
+                getInitialTokens()
+            } catch (e: Exception) {
+                logger.error { e.message }
```
Comment (Deseteral) [inline service/src/main/kotlin/xyz/lebkuchenfm/external/storage/dropbox/DropboxClient.kt:59]:
```suggestion
                logger.error(e) { "Could not get Dropbox auth tokens." }
```

File: service/src/main/kotlin/xyz/lebkuchenfm/external/storage/dropbox/DropboxClient.kt [hunk 5]
```diff
@@ -29,21 +32,33 @@ import xyz.lebkuchenfm.external.storage.dropbox.models.DropboxFileUploadArgs
...
 
     /**
      * [path] must contain destination folder, file name and its extension
      * [bytes] contains file content data
      * @return url on which raw file may be obtained
      */
-    suspend fun uploadFile(path: String, bytes: ByteArray): String {
+    suspend fun uploadFile(path: String, bytes: ByteArray): Result<String, DropboxClientError> {
+        if (refreshToken == null || appKey == null || appSecret == null) {
+            logger.warn { "Tried to use dropbox client but config is missing." }
```
Comment (Deseteral) [inline service/src/main/kotlin/xyz/lebkuchenfm/external/storage/dropbox/DropboxClient.kt:51]:
```suggestion
            logger.warn { "Tried to use Dropbox client but its config is missing." }
```

File: service/src/main/kotlin/xyz/lebkuchenfm/domain/xsounds/XSoundsFileRepository.kt [hunk 6]
```diff
@@ -1,5 +1,11 @@
 package xyz.lebkuchenfm.domain.xsounds
 
+import com.github.michaelbull.result.Result
+
 interface XSoundsFileRepository {
-    suspend fun uploadXSoundFile(soundName: String, byteArray: ByteArray): String
+    suspend fun uploadXSoundFile(soundName: String, byteArray: ByteArray): Result<String, FileStorageError>
+}
+
+sealed interface FileStorageError {
```
Comment (Deseteral) [inline service/src/main/kotlin/xyz/lebkuchenfm/domain/xsounds/XSoundsFileRepository.kt:9]: I feel like `FileStorageError` is too generic in this case. This should be about uploading new sound file.
```suggestion
sealed interface UploadXSoundFileError {
```

**Review by Deseteral — CHANGES_REQUESTED** [review]

File: service/src/main/kotlin/xyz/lebkuchenfm/api/xsounds/XSoundsEndpoint.kt (hunk 1 above)
Comment (gorzalczany) [reply service/src/main/kotlin/xyz/lebkuchenfm/api/xsounds/XSoundsEndpoint.kt:64]: added to Doc

File: service/src/main/kotlin/xyz/lebkuchenfm/domain/xsounds/XSoundsService.kt [hunk 7]
```diff
@@ -19,15 +19,30 @@ class XSoundsService(private val repository: XSoundsRepository, private val file
     }
 
     sealed interface NewSoundError {
-        data object CouldNotAddNewSound : NewSoundError
+        data object FileStorageError : NewSoundError
+        data object DataBaseError : NewSoundError
```
Comment (Deseteral) [inline service/src/main/kotlin/xyz/lebkuchenfm/domain/xsounds/XSoundsService.kt:23]: Database is a single word 😉 
```suggestion
        data object DatabaseError : NewSoundError
```

File: service/src/main/kotlin/xyz/lebkuchenfm/external/storage/dropbox/models/DropboxFileUpload.kt [hunk 8]
```diff
@@ -9,6 +9,7 @@ data class DropboxFileUploadArgs(
     val mode: String,
     val autorename: Boolean,
     val mute: Boolean,
+    @SerialName("strict_conflict") val strictConflict: Boolean,
```
Comment (Deseteral) [inline service/src/main/kotlin/xyz/lebkuchenfm/external/storage/dropbox/models/DropboxFileUpload.kt:12]: Wait, why `@SerialName` here? What does it do? Rename from snake_case to camelCase?

Comment (gorzalczany) [reply service/src/main/kotlin/xyz/lebkuchenfm/external/storage/dropbox/models/DropboxFileUpload.kt:12]: Yup. Formatter (inteliJ?) complains at snake case but thats how it is in api.

Comment (Deseteral) [reply service/src/main/kotlin/xyz/lebkuchenfm/external/storage/dropbox/models/DropboxFileUpload.kt:12]: Funny that it's the same annotation for field rename and type discrimination.

**Review by Deseteral — APPROVED** [review]

**Pull request merged**